atoma-vllm/src/worker.rs

```diff
@@ -0,0 +1,690 @@
+        })
+    }
+
+    /// Determines the number of available GPU blocks
+    pub fn num_available_gpu_blocks(&self) -> usize {
+        todo!()
+    }
+
+    /// Executes model's forward pass
+    #[instrument(skip_all)]
+    pub fn execute_model(
+        &mut self,
+        request: ExecuteModelRequest,
+    ) -> Result<Vec<SequenceGroupOutput>, ModelWorkerError> {
+        info!("Executing model on new request..");
+
+        let span = self.span.clone();
+        let _enter = span.enter();
+
+        let ExecuteModelRequest {
+            sequence_groups_metadata,
+            blocks_to_swap_in,
+            blocks_to_swap_out,
+            blocks_to_copy,
+            ..
+        } = request;
+
+        let num_sequence_groups = sequence_groups_metadata.len();
+
+        // `blocks_to_copy` is a GPU tensor. The source and target of
+        // blocks to copy are in the same device, and `blocks_to_copy`
+        // can be used directly within cuda kernels.
+        let blocks_to_copy = blocks_to_copy
+            .into_iter()
+            .flat_map(|(i, j)| [i, j])
+            .collect::<Vec<_>>();
+        let blocks_to_copy = if blocks_to_copy.is_empty() {
+            None
+        } else {
+            let num_block_pairs_to_copy = blocks_to_copy.len() / 2;
+            Some(Tensor::from_vec(
+                blocks_to_copy,
+                (num_block_pairs_to_copy, 2),
+                &self.device,
+            )?)
+        };
+
+        // At this point we need to perform cache swap operations
+        self.cache_swap(&blocks_to_swap_in, &blocks_to_swap_out, blocks_to_copy)?;
+
+        // NOTE: Number of sequence groups should not be zero,
+        // as we don't schedule empty sequences, for now.
+        if num_sequence_groups == 0 {
+            warn!("Number of sequence groups to run model on should not be empty");
+            return Ok(vec![]);
+        }
+
+        let ModelInput {
+            input_tokens_tensor,
+            input_positions,
+            attention_metadata,
+            cu_query_lengths,
+            ..
+        } = self.prepare_input_tensors(&sequence_groups_metadata)?;
+
+        let selected_token_indices = utils::compute_selected_token_indices(&cu_query_lengths)?;
+
+        let kv_cache = self.cache_engine.gpu_cache.iter_mut().collect();
+        let logits = self.model.forward(
+            &input_tokens_tensor,
+            &input_positions,
+            &selected_token_indices,
+            kv_cache,
+            attention_metadata,
+        )?;
+
+        let sampled_outputs = self.model.sample(&logits, &sequence_groups_metadata)?;
+
+        Ok(sampled_outputs)
+    }
+
+    /// Swaps cached blocks
+    #[instrument(skip_all)]
+    pub fn cache_swap(
+        &mut self,
+        blocks_to_swap_in: &HashMap<i64, i64>,
+        blocks_to_swap_out: &HashMap<i64, i64>,
+        blocks_to_copy: Option<Tensor>,
+    ) -> Result<(), ModelWorkerError> {
+        if blocks_to_swap_in.len() > 0 {
```

Comment:
When the memory is getting reorganized it usually means it's almost full, right? That's why we need to swap out some blocks etc. The question is, is it safe to first swap in (as the gpu can be full at this point). Shouldn't we first swap out the blocks from the gpu?

Reply:
Does it really matter ? Both the CPU/GPU memory is allocated at initialization, with a zeroed `KV` cache. This means that whenever we `swap` operations, we are not necessarily allocating new memory, we are just writing from CPU<->GPU. Once the algorithm chooses what has to swap in/out, these operations should be quite independent.. 

What do you think ?

Reply:
Swapping it means moving from cpu to gpu, right? What if all you gpu blocks are used? There is no place to swap them into.